### ISSUE TITLE
[Question] Can nvitop keep a log/record of GPU-Utilization and store in a CSV?

### ISSUE THREAD

**Issue by FelixMildon** [XuehaiPan/nvitop#20]
I'm trying to record GPU-utilization, and the users., and what programs are running.
Is there a way to log and save this information? Like into a CSV or database?

Sorry if I missed something from the readme

**Comment by XuehaiPan** [comment]
@FelixMildon Hi, `nvitop` provides APIs to query the device and process status, see https://github.com/XuehaiPan/nvitop#device.

```python
from nvitop import Device, GpuProcess

all_devices = Device.all()
all_gpu_processes = []
for device in all_devices:
    all_gpu_processes.extend(device.processes().values())

all_device_snapshots = [device.as_snapshot() for device in all_devices]
all_gpu_process_snapshots = [process.as_snapshot() for process in all_processes]
```

I'm going to add `tensorboard` integration this weekend (see https://github.com/XuehaiPan/nvitop/issues/18#issuecomment-1140715085), which will introduce more user-friendly interfaces.

**Comment by FelixMildon** [comment]
@XuehaiPan I'm really unsure how to use NVItop to log processes and gpu-utalisation (from nvidia-smi). Logging these in intervals into a CSV would be really useful for University of Aucklands GPUs.

**Comment by XuehaiPan** [comment]
Hi @FelixMildon, I implement a new metric collector in PR #21, which can collect the resource consumption for GPUs and processes.

```python
import datetime
import time

import pandas as pd

from nvitop import ResourceMetricCollector

collector = ResourceMetricCollector(root_pids={1}, interval=2.0)  # log all devices and all GPU processes
df = pd.DataFrame()

with collector(tag='resources'):
    for _ in range(60):
        # Do something
        time.sleep(60)

        metrics = collector.collect()
        df_metrics = pd.DataFrame.from_records(metrics, index=[len(df)])
        df = pd.concat([df, df_metrics], ignore_index=True)
        # Flush to CSV file ...

df.insert(0, 'time', df['resources/timestamp'].map(datetime.datetime.fromtimestamp))
df.to_csv('results.csv', index=False)
```

See also #18.

**Comment by XuehaiPan** [comment]
Close as resolved by PR #21.
